SIGSEGV in fetchRequest.AppendTo, might be related to ConsumePreferringLagFn

Hi Travis, 

We are experiencing panics in our consumer right now. It _seems_ only to be an issue when `ConsumePreferringLagFn` is being used. We had the consumers offline for an hour, when we redeployed this started. Please advice.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xbd9f4b]

goroutine 25322 [running]:
github.com/twmb/franz-go/pkg/kgo.(*fetchRequest).AppendTo(0xc0fca9c3c0, {0xc00e5c0800, 0xf, 0x400})
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.0/pkg/kgo/source.go:1815 +0x8ab
github.com/twmb/franz-go/pkg/kmsg.(*RequestFormatter).AppendRequest(0xc000466b10, {0xc00e5c0800?, 0x19249c0?, 0x0?}, {0x1204f20, 0xc0fca9c3c0}, 0x4)
	/go/pkg/mod/github.com/twmb/franz-go/pkg/kmsg@v1.2.0/api.go:233 +0x3c9
github.com/twmb/franz-go/pkg/kgo.(*brokerCxn).writeRequest(0xc002c00400, {0x11ffc40, 0xc001de50c0}, {0x1204f20?, 0xc0fca9c3c0?, 0x19249c0?}, {0x1204f20, 0xc0fca9c3c0})
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.0/pkg/kgo/broker.go:1006 +0x425
github.com/twmb/franz-go/pkg/kgo.(*broker).handleReq(0xc0005cdb80, {{0x11ffc40, 0xc001de50c0}, {0x1204f20, 0xc0fca9c3c0}, 0xc001980540, {0xc0ea180e93b3bd4f, 0x395ea367b, 0x19249c0}})
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.0/pkg/kgo/broker.go:428 +0xa8d
github.com/twmb/franz-go/pkg/kgo.(*broker).handleReqs(0xc0005cdb80, {{0x11ffc40, 0xc001de50c0}, {0x1204f20, 0xc0fca9c3c0}, 0xc001980540, {0xc0ea180e93b3bd4f, 0x395ea367b, 0x19249c0}})
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.0/pkg/kgo/broker.go:273 +0xbd
created by github.com/twmb/franz-go/pkg/kgo.(*broker).do
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.0/pkg/kgo/broker.go:247 +0x229
```

Are there any information we can provide that will help you debug these panics?

CC: @jepp2078

--- a/pkg/kgo/group_test.go
+++ b/pkg/kgo/group_test.go
@@ -124,7 +124,7 @@ func (c *testConsumer) etl(etlsBeforeQuit int) {
 		ConsumeTopics(c.consumeFrom),
 		Balancers(c.balancer),
 		MaxBufferedRecords(10000),
-		ConsumePreferringLagFn(PreferLagAt(10)),
+		ConsumePreferringLagFn(PreferLagAt(1)),
 
 		// Even with autocommitting, autocommitting does not commit
 		// *the latest* when being revoked. We always want to commit

--- a/pkg/kgo/helpers_test.go
+++ b/pkg/kgo/helpers_test.go
@@ -12,6 +12,7 @@ import (
 	"strconv"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -34,6 +35,12 @@ var (
 	// EndTxn to return successfully before beginning a new transaction. We
 	// cannot use EndAndBeginTransaction with EndBeginTxnUnsafe.
 	allowUnsafe = false
+
+	// We create topics with a different number of partitions to exercise
+	// a few extra code paths; we index into npartitions with npartitionsAt,
+	// an atomic that we modulo after load.
+	npartitions   = []int{7, 11, 31}
+	npartitionsAt int64
 )
 
 func init() {
@@ -110,10 +117,11 @@ func tmpTopic(tb testing.TB) (string, func()) {
 
 	topic := randsha()
 
+	partitions := npartitions[int(atomic.AddInt64(&npartitionsAt, 1))%len(npartitions)]
 	req := kmsg.NewPtrCreateTopicsRequest()
 	reqTopic := kmsg.NewCreateTopicsRequestTopic()
 	reqTopic.Topic = topic
-	reqTopic.NumPartitions = 20
+	reqTopic.NumPartitions = int32(partitions)
 	reqTopic.ReplicationFactor = int16(testrf)
 	req.Topics = append(req.Topics, reqTopic)
 

--- a/pkg/kgo/sink.go
+++ b/pkg/kgo/sink.go
@@ -536,9 +536,9 @@ func (s *sink) handleReqClientErr(req *produceRequest, err error) {
 		isRetriableBrokerErr(err):
 		updateMeta := !isRetriableBrokerErr(err)
 		if updateMeta {
-			s.cl.cfg.logger.Log(LogLevelInfo, "produce request failed triggering metadata update", "broker", logID(s.nodeID), "err", err)
+			s.cl.cfg.logger.Log(LogLevelInfo, "produce request failed, triggering metadata update", "broker", logID(s.nodeID), "err", err)
 		}
-		s.handleRetryBatches(req.batches, req.backoffSeq, updateMeta, false, "failed produce request triggering metadata update")
+		s.handleRetryBatches(req.batches, req.backoffSeq, updateMeta, false, "failed produce request triggered metadata update")
 
 	case errors.Is(err, ErrClientClosed):
 		s.cl.failBufferedRecords(ErrClientClosed)

--- a/pkg/kgo/source.go
+++ b/pkg/kgo/source.go
@@ -1582,13 +1582,17 @@ func (f *fetchRequest) addCursor(c *cursor) {
 // order, and the previously determined per-topic partition order, and returns
 // a new topic and per-topic partition order.
 //
-// Most use cases will not need to look at the prior orders, but they exist
-// if you want to get fancy.
+// Most use cases will not need to look at the prior orders, but they exist if
+// you want to get fancy.
 //
 // You can return partial results: if you only return topics, partitions within
 // each topic keep their prior ordering. If you only return some topics but not
 // all, the topics you do not return / the partitions you do not return will
 // retain their original ordering *after* your given ordering.
+//
+// NOTE: torderPrior and porderPrior must not be modified. To avoid a bit of
+// unnecessary allocations, these arguments are views into data that is used to
+// build a fetch request.
 type PreferLagFn func(lag map[string]map[int32]int64, torderPrior []string, porderPrior map[string][]int32) ([]string, map[string][]int32)
 
 // PreferLagAt is a simple PreferLagFn that orders the largest lag first, for
@@ -1704,6 +1708,9 @@ func (f *fetchRequest) adjustPreferringLag() {
 			if c.offset <= 0 {
 				lag = hwm
 			}
+			if lag < 0 {
+				lag = 0
+			}
 			plag[p] = lag
 		}
 	}
@@ -1712,29 +1719,34 @@ func (f *fetchRequest) adjustPreferringLag() {
 	if torder == nil && porder == nil {
 		return
 	}
-	if torder == nil {
-		torder = f.torder
-	}
-	if porder == nil {
-		porder = f.porder
-	}
 	defer func() { f.torder, f.porder = torder, porder }()
 
-	// Remove any extra topics the user returned that we were not
-	// consuming, and add all topics they did not give back.
-	for i := 0; i < len(torder); i++ {
-		t := torder[i]
-		if _, exists := tall[t]; !exists {
-			torder = append(torder[:i], torder[i+1:]...) // user gave topic we were not fetching
-			i--
+	if len(torder) == 0 {
+		torder = f.torder // user did not modify topic order, keep old order
+	} else {
+		// Remove any extra topics the user returned that we were not
+		// consuming, and add all topics they did not give back.
+		for i := 0; i < len(torder); i++ {
+			t := torder[i]
+			if _, exists := tall[t]; !exists {
+				torder = append(torder[:i], torder[i+1:]...) // user gave topic we were not fetching
+				i--
+			}
+			delete(tall, t)
 		}
-		delete(tall, t)
-	}
-	for t := range tall {
-		torder = append(torder, t) // user did not return topic we were fetching
+		for _, t := range f.torder {
+			if _, exists := tall[t]; exists {
+				torder = append(torder, t) // user did not return topic we were fetching
+				delete(tall, t)
+			}
+		}
 	}
 
-	// Now, same thing for partitions.
+	if len(porder) == 0 {
+		porder = f.porder // user did not modify partition order, keep old order
+		return
+	}
+
 	pused := make(map[int32]struct{})
 	for t, ps := range pall {
 		order, exists := porder[t]
@@ -1753,8 +1765,11 @@ func (f *fetchRequest) adjustPreferringLag() {
 			}
 			delete(pused, p)
 		}
-		for p := range pused {
-			order = append(order, p)
+		for _, p := range f.porder[t] {
+			if _, exists := pused[p]; exists {
+				order = append(order, p)
+				delete(pused, p)
+			}
 		}
 		porder[t] = order
 	}

